src/Xrefcheck/System.hs

```diff
@@ -36,21 +42,67 @@ readingSystem = unsafePerformIO
 -- Check the respective env variable which is usually set in all CIs.
 askWithinCI :: IO Bool
 askWithinCI = lookupEnv "CI" <&> \case
-  Just "1"                       -> True
+  Just "1" -> True
   Just (map C.toLower -> "true") -> True
-  _                              -> False
+  _ -> False
+
+-- | A FilePath that has been canonicalized with 'System.Directory.canonicalizePath',
+-- and with indirections expanded as much as possible. Should be created via
+-- 'mkGlobPattern'.
+newtype CanonicalPath = UnsafeCanonicalPath
+  { unCanonicalPath :: FilePath
+  } deriving newtype (Show, Eq, Ord)
+
+canonicalizePath :: FilePath -> IO CanonicalPath
+canonicalizePath = (UnsafeCanonicalPath . expandIndirections <$>) . Directory.canonicalizePath
+  where
+  -- 'canonicalizePath' does not handle indirections when the file does not exist,
+  -- but we can expand as much as we can syntactically.
+  expandIndirections :: FilePath -> FilePath
+  expandIndirections = Posix.joinPath . reverse . expand 0 . reverse . Posix.splitDirectories
+
+  expand :: Int -> [FilePath] -> [FilePath]
+  expand acc (".." : xs) = expand (acc + 1) xs
+  expand acc ("." : xs) = expand acc xs
+  expand 0 (x : xs) = x : expand 0 xs
+  expand acc (_ : xs) = expand (acc - 1) xs
+  expand acc [] = replicate acc ".."
 
--- | Glob pattern relative to repository root. Should be created via @mkGlobPattern@
-newtype RelGlobPattern = RelGlobPattern FilePath
+takeDirectory :: CanonicalPath -> CanonicalPath
+takeDirectory (UnsafeCanonicalPath p) = UnsafeCanonicalPath $ Posix.takeDirectory p
+
+getDirsBetweenRootAndFile :: CanonicalPath -> CanonicalPath -> [CanonicalPath]
+getDirsBetweenRootAndFile root@(UnsafeCanonicalPath rootPath) file =
+  case getRelativeChild root (takeDirectory file) of
+    Just path -> UnsafeCanonicalPath
+      <$> scanl (Posix.</>) rootPath (Posix.splitDirectories path)
+    Nothing -> []
+
+getRelativeChild :: CanonicalPath -> CanonicalPath -> Maybe FilePath
+getRelativeChild (UnsafeCanonicalPath root) (UnsafeCanonicalPath child) =
+  dropWhile Posix.isPathSeparator <$> stripPrefix root child
```

Comment:
Note here that we seem to need to use a cross-platform version (from `System.FilePath`), not the Posix one, since the canonicalized path already contains `\` on windows.

I suppose this is a problem coming from the previous fixes, we used a big hammer and switched to `System.FilePath.Posix` more or less everywhere in the project, because markdown links are always separated with Posix'es `/` (when they point to a local place). This is an unexpected layer of problems, but now it may be the perfect to refactor this.

Ideally, to put a clear bound on where we expect a Posix-formatted filepath and where we expect a platform-dependent filepath, I think it would be cool to add another newtype for markdown links, and clarify that they are always like in Unix systems.

Then carefully pass through all our filepath-related utilities and update them respectively.

Does it make sense?

Reply:
I'm actually not 100% sure that I'm correct in my understanding, but we must end up with tests covering this logic, and I hope they will reveal what is correct.

Reply:
That makes sense! I was wondering why the Posix version was being used. I will try to use the Posix version only with the code that is related to Markdown links.

Reply:
Good :+1: :+1: